### PR TITLE
Add comparator option

### PR DESCRIPTION
**Pull request by HollyPony**
Allow manual comparison manipulating arrays

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -96,23 +96,32 @@ module.exports = {
...
-          added = _.differenceWith(val.$appendUnique, array, _.isEqual);
-          return array.concat(added);
+          array = (_.get(moduleOptions, sliced) || []).concat(val.$appendUnique);
+          return _.reverse(_.uniqWith(_.reverse(array), comparator));
         } else if (val.$prependUnique) {
-          array = _.get(moduleOptions, sliced) || [];
-          added = _.differenceWith(val.$prependUnique, array, _.isEqual);
-          return added.concat(array || []);
+          array =  val.$prependUnique.concat(_.get(moduleOptions, sliced) || []);
+          return _.uniqWith(array, comparator);
```
Comment (HollyPony) [inline index.js]: @boutell  This is a "breaking" change, assuming the value I want to `prepend` will be unique on the head of the array. Assuming the user want to force move a value in the array.

Comment (boutell) [reply index.js]: Forcing the move seems incorrect to me, also this code would throw out all duplicates, not just duplicates that are relevant to the list in prependUnique or appendUnique.

**Review by boutell — CHANGES_REQUESTED** [review]
Please take a look at my comment — I'm not sure forcing the move is correct, but more importantly this code would discard duplicates that have nothing to do with the list that was passed.

**Comment by HollyPony** [comment]
2 things:
1: It was my expectation writing tests. if I have a base `[1, 2, 3]` and I want to prepend 3, I assume 3 will be the first (resulting to `[3, 1, 2]`). Because the "pretender" has right. It's related to point 2 ...

2: ... that's what we need, in case of I've no Integers but Objects. Each object is an analytic key/value tag. I want to avoid any duplicates entries, our use cases work with key/value corresponding to analytics elements. So, we need the comparator on "keys". And the element added should be the one added. And that's doesn't work with differenceWith.

As I understand by `code would discard duplicates that have nothing to do with the list`, you mean, with a base of `[1, 2, 2, 2, 2, 3]` if I `appendUnique: [3]`, the result is untouched ? If yes, I should update test cases.

EDIT: If I append `[2, 2]` to the list above, the result should be `[1, 3, 2, 2]` or `[1, 3, 2]` ? (I vote for the second solution)

Could you provide me a use case to counter mines ?

**Comment by boutell** [comment]
Arthur, technically we've already npm published this as a semver-stable module, so we shouldn't be changing the semantics of these operators. But since they are absolutely brand new, I think we are safe in believing no one else is already relying on it to behave the way I had in mind.

So, we can go ahead and say that if you use the Unique operators, they will always make sure the element is last, or first, even if it was already present.

However, I still do not think they should remove duplicates of things that aren't being prepended or appended. That is a side effect that is not obvious and not our concern here. Some things make sense more than once and others do not.  If everything should be unique then just use these operators consistently and the situation will not arise.

**Comment by HollyPony** [comment]
You right, I'll make another proposal around a new `$replace` operator that's match with our needs and does not break semantic behaviour.

**Comment by HollyPony** [comment]
It should be ok for now.

File: index.js [hunk 2]
```diff
@@ -104,16 +115,26 @@ module.exports = {
...
+          added = _.differenceWith(val.$prependUnique, array, comparator);
           return added.concat(array || []);
         } else if (val.$remove) {
           array = _.get(moduleOptions, sliced);
-          array = _.differenceWith(array, val.$remove, _.isEqual);
+          array = _.differenceWith(array, val.$remove, comparator);
           return array;
+        } else if (val.$replace) {
+          if (!val.comparator) {
+            console.warn('Using \'$replace\' without \'comparator\' is probably a bug')
```
Comment (boutell) [inline index.js]: Missing semicolon.

**Review by boutell — CHANGES_REQUESTED** [review]
Just one really minor style issue then I'll merge. Thanks for the feature and the tests.

**Comment by HollyPony** [comment]
@boutell Done !

**Review by boutell — APPROVED** [review]

**Comment by boutell** [comment]
This has been published to npm.

**Pull request merged**